vscode-dotnet-runtime-library/src/Utils/NodeIPCMutex.ts

```diff
@@ -0,0 +1,286 @@
+        // In Linux, The user or system may set XDG_RUNTIME_DIR to set our applications temporary directory, so this this instead of /temp/
+        // On Unix, Access to /tmp/ may be restricted, but all processes must use the same directory, so we can't condition to use another dir based on the permissions for it.
+
+        // On Windows, '\\\\.\\pipe\\` is a Special File System to get a Named Pipe (File Descriptors won't work)
+        // https://nodejs.org/docs/latest/api/net.html#ipc-support:~:text=On%20Windows%2C%20the,owning%20process%20exits.
+
+        const ipcPathDir = os.platform() === 'win32' ? `\\\\.\\pipe\\` :
+            os.platform() === 'linux' && process.env.XDG_RUNTIME_DIR ? process.env.XDG_RUNTIME_DIR as string : os.tmpdir();
+
+        if (id.length > (lengthLimit - ipcPathDir.length))
+        {
+            this.logger.log(`Lock ID is too long, truncating to ${lengthLimit - ipcPathDir.length} characters, due to dir: ${ipcPathDir}`);
+            id = id.substring(0, Math.max(lengthLimit - 1, 6)); // Prevent setting the environment variable to make the application fail. the vscd prefix + 2 should uniquely identify most locks.
+        }
+
+        if (process.platform === 'win32')
+        {
+            return `${ipcPathDir}vscd-${id}-sock`;
+        }
+        else
+        {
+            return path.join(ipcPathDir, `vscd-${id}.sock`);
+        }
+    }
+
+    /**
+     * @remarks This function will try to hold a lock to prevent both other processes and other async code in this processes from running simultaneously..
+     * It will retry acquiring the lock if it is already held by another process or async code in this process.
+     * It will also check if the lock is stale (i.e., if the process holding the lock has died) and clean it up if necessary.
+     * It will fail and throw if it cannot acquire the lock after the specified number of retries.
+     *
+     * @param fn - The function to run while we have the lock. This should be a promise, and will be awaited with its value returned.
+     * Use the () => {} syntax to ensure your scope is correctly bound to the function so it can access the variables you declare.
+     * @param retryDelayMs The number of milliseconds to wait before retrying to acquire the lock if it is already held by another process.
+     * @param timeoutTimeMs The total amount of time to try to acquire the lock before giving up. This is the maximum time to wait for the lock to be released.
+     * @param actionId The action ID to use for logging and debugging purposes. This should be a unique identifier for the action being performed.
+     * @returns The awaited value returned by the function passed in as fn.
+     */
+    public async acquire<T>(fn: () => Promise<T>, retryDelayMs = 100, timeoutTimeMs = 1000, actionId: string): Promise<T>
+    {
+        const maxRetryCountToEndAtRoughlyTimeoutTime = Math.ceil(timeoutTimeMs / retryDelayMs) + 1;
```

Comment:
`maxRetryCountToEndAtRoughlyTimeoutTime` 😅

Reply:
My variable name here is definitely amusing 😄 
It doesn't exactly end at the timeout time since it's not a time-based timeout, and instead a number of retries based timeout. I could improve this, though I would rather not for the scope of this PR.

Reply:
Note: this could theoretically be a bigger issue, but all of the callers to this are already wrapped in their own custom timeouts with timeout logic.